Disconnect velux on hass stop

gibman opened the pull request:
## Description:

Often when rebooting hass the KLF 200 device would not disconnect the previous connection which in turns locks out the use from connecting again.
In these situations one would have to powercycle the klf200 velux device, which could be a problem if not being at home.

Ive set up af listener for EVENT_HOMEASSISTANT_STOP, which would jsutm ake sure to call disconnect on the device.

more info here:
https://community.home-assistant.io/t/velux-component-for-klf-200-doesnt-support-the-new-api-with-firmware-2-0-0-71/75641/67

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

gibman committed:
velux KLF200 device did not disconnect properly when rebooting the hass device.
disconnect is now being called on the 'EVENT_HOMEASSISTANT_STOP' event

gibman committed:
removed comment

gibman committed:
removed comment

homeassistant commented:
Hello @gibman,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`1f6fb5db5af81d28308d02d15a5d5bf867badeaf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;26266&#x2F;commits&#x2F;1f6fb5db5af81d28308d02d15a5d5bf867badeaf) - This commit has something that looks like an email address (elr@seges.dk). Maybe try linking that to GitHub?.
- [`326e3c0915bc3ebf53727ca5627538217eb30907`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;26266&#x2F;commits&#x2F;326e3c0915bc3ebf53727ca5627538217eb30907) - This commit has something that looks like an email address (elr@seges.dk). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

gibman committed:
trigger bot

gibman committed:
trigger bot

gibman committed:
Merge branch 'master' of https://github.com/gibman/home-assistant

gibman committed:
trigger bot

gibman commented:
So Ive added the "unknown" email to my account .. how do I retrigger the bot in order to re-evaluate?
I tried doing an empty commit along with a dummy comment.
Doesnt seem to cut it...

sigh..

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -40,14 +40,19 @@
+        
+        async def on_hass_stop(event):
+            """Close connection when hass stops."""
+            _LOGGER.info("velux interface terminated")
```
Please start logging messages with capital letter.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -40,14 +40,19 @@
+            _LOGGER.info("velux interface terminated")
+            await self.pyvlx.disconnect()                    
+        
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)                   
```
We try to avoid side effects in init method. Please extract all the side effects into a dedicated setup method.

gibman committed:
logging casing fixed. code moved from init.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -40,24 +41,35 @@
         from pyvlx import PyVLX
-        host = config[DOMAIN].get(CONF_HOST)
-        password = config[DOMAIN].get(CONF_PASSWORD)
+        host = self._config[DOMAIN].get(CONF_HOST)
```
Please only pass the domain part of the config to this instance.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
+            await self.pyvlx.disconnect()
+            
+        self._hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)     
+        from pyvlx import PyVLX
```
Please move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
-        from pyvlx import PyVLX
+        self._hass = hass
+        self._config = config
+        
```
Initialize `self.pyvlx` to `None`. All instance attributes must be initialized in init method.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
+
     async def async_start(self):
         """Start velux component."""
+        _LOGGER.info("Velux interface started")        
```
Debug level max.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
         await self.pyvlx.load_nodes()
+                        
+        for node in self.pyvlx.nodes:
+          _LOGGER.info("Stopping node: " + str(node))
```
Debug level max.

```py
_LOGGER.debug("Stopping node: %s", node)
```

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
+    def setup(self):
+        async def on_hass_stop(event):
+            """Close connection when hass stops."""
+            _LOGGER.info("Velux interface terminated")
```
Debug max.

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -46,13 +47,29 @@ class VeluxModule:
+                        
+        for node in self.pyvlx.nodes:
+          _LOGGER.info("Stopping node: " + str(node))
+          await node.stop(wait_for_completion=False)
```
Why do we need to stop the nodes here?

gibman replied:
I'm sorry.. It seemed to sneak its way in there as its part of another fix that does not belong to this particular PR.

should be removed in latest commit.

gibman committed:
logger level debug

logger level moved from info to debug
only config[DOMAIN] exposed to module
imports moved to top

MartinHjelmare commented on homeassistant/components/velux/__init__.py:
```diff
@@ -23,9 +24,7 @@
-
+    """Set up the velux component."""    
     try:
         hass.data[DATA_VELUX] = VeluxModule(hass, config)
```
Please just pass `config[DOMAIN]` here. The `VeluxModule` class doesn't need to know about the whole config.

gibman committed:
DOMAIN part of config passed to module.

MartinHjelmare reviewed (approved):
Good! Can be merged when lint and formatting is fixed and build passes.

gibman commented:
> Good! Can be merged when lint and formatting is fixed and build passes.

No idea what to fix here  ??
Not too many hints revealed in those error messages :/

ppetru commented:
@gibman thanks for the PR, looking forward to having it in a release since it's quite annoying to keep powercycling my KLF 200 :)

I'm not too familiar with the automated checks involved here, but clicking through the details links brought me to this: https://dev.azure.com/home-assistant/Home%20Assistant/_build/results?buildId=8149&view=logs&jobId=2fa2c639-d05e-530d-034e-4fb413fcea4e&j=4e214fed-f310-53fa-0018-b38e7beb405b -- if you expand the "Run flake8" section you get a bunch of lint errors:

homeassistant/components/velux/__init__.py:27:38: W291 trailing whitespace
homeassistant/components/velux/__init__.py:50:26: W291 trailing whitespace
homeassistant/components/velux/__init__.py:52:1: W293 blank line contains whitespace
homeassistant/components/velux/__init__.py:54:1: D102 Missing docstring in public method
homeassistant/components/velux/__init__.py:54:5: E303 too many blank lines (2)
homeassistant/components/velux/__init__.py:59:1: W293 blank line contains whitespace
homeassistant/components/velux/__init__.py:60:81: W291 trailing whitespace
homeassistant/components/velux/__init__.py:66:5: E303 too many blank lines (2)
homeassistant/components/velux/__init__.py:68:49: W291 trailing whitespace
homeassistant/components/velux/__init__.py:71:1: W391 blank line at end of file
homeassistant/components/velux/__init__.py:71:1: W293 blank line contains whitespace

This might also be useful for checking locally so that you don't have to wait for the CI robots to point out your errors: https://developers.home-assistant.io/docs/en/development_testing.html#preventing-linter-errors

gibman committed:
removed trailing whitespaces etc.

MartinHjelmare commented:
Please run black from the project root and commit the changes, to solve formatting.
```
black --fast homeassistant
```

gibman commented:
> Please run black from the project root and commit the changes, to solve formatting.
> 
> ```
> black --fast homeassistant
> ```

I think you assume that I know what black is ? :)
I'm merely running  notepad++ on a windows box.
I did try to get rid of those annoying trailing whitespaces and the latest commit

regards
gibman

ppetru commented:
This might be helpful: https://black.readthedocs.io/en/stable/installation_and_usage.html

gibman committed:
black --fast changes

gibman committed:
added missing docstring

MartinHjelmare commented:
Black is mentioned in our developer docs:
https://developers.home-assistant.io/docs/en/development_guidelines.html

gibman committed:
D400 First line should end with a period

gibman committed:
black formatting

MartinHjelmare commented:
Thanks!
